Virtual Floor

JeroenDStout opened the pull request:
Add a virtual floor to the game, activated by the alt/shift method of placing things at different heights.

![surface0](https://user-images.githubusercontent.com/14242454/31050274-e3f0e62e-a645-11e7-9b10-ec26b733d631.gif)

It displays a grid on which obstructing elements and walls are highlighted as bright edges.

![surface1](https://user-images.githubusercontent.com/14242454/31050275-e41e9006-a645-11e7-9b6d-e6ae6212fb1a.gif)

While constructing a ride pressing shift places the virtual floor on a corresponding level, making it possible to see the possible scenery collisions.

![surface2](https://user-images.githubusercontent.com/14242454/31050276-e420c45c-a645-11e7-9250-c20b2d3f8f1f.gif)

It also works underground.

![underground](https://user-images.githubusercontent.com/14242454/31050297-8215b0dc-a646-11e7-863b-fbbfceba8503.jpg)

This is WIP:
- Currently only works in OpenGL renderer (redraw regions still need to be added)
- It is not yet fully updated for constructing paths
- It should probably have a toggle in view options
- I believe it could be tweaked to be better, based on feedback. There potentially are also some bugs which my own play-style has hidden from me.

But:
- Some sprites will never proper sort with the grid unless the sorting is fixed.

JeroenDStout commented:
Could someone help me understand the Travis errors?

I am not sure what `ninja: build stopped: cannot make progress due to previous errors` (no further information) means. The other error is just `ld: symbol(s) not found for architecture x86_64`, which I also do not understand.

janisozaur commented:
@JeroenDStout like ninja ssays, the error is a bit earlier. See https://travis-ci.org/OpenRCT2/OpenRCT2/jobs/281783225#L589:

> map_element.c:(.text+0x77e): undefined reference to `virtual_floor_paint'

Most likely difference in bindings (C vs C++) in some place.

And don't worry about Xcode, it's special snowflake and needs to be dealt with appropriately.

janisozaur commented on src/openrct2/world/map.c:
```diff
@@ -4624,3 +4628,17 @@ uint16 check_max_allowable_land_rights_for_tile(uint8 x, uint8 y, uint8 base_z)
+
+void map_invalidate_virtual_floor_tiles()
+{
+}
```
missing newline

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,286 @@
+        sub_98197C(session, SPR_G2_SELECTION_EDGE_SE | (!(occupiedEdges & 0x2)? ((litEdges & 0x2)? remap_lit : remap_base) : remap_edge),
+            0, 0, 1, 1, 1, gMapVirtualFloorHeight, 16, 27, gMapVirtualFloorHeight + ((dullEdges & 0x2)? -2 : 0), get_current_rotation());
+    }
+}
```
missing newline

janisozaur commented:
Ride supports don't cause the edges to become white. They look very solid to me and I expected them to be identified as collision too.

janisozaur commented:
The new sprites are not optimised.

PFCKrutonium commented:
@janisozaur Ride Supports don't cause collisions - they move out of the way or are removed entirely depending on what is under them. I support them not being marked as collision, since they aren't.

JeroenDStout commented:
@janisozaur @PFCKrutonium The purpose is primarily a collision test, in the end. Supports could be added but would give you a false impression of the region; unless they were a different colour, but then the amount of information may just become too large.

JeroenDStout commented:
@janisozaur Could you help out with the `virtual_floor_paint` problem? As far as I can tell, `map_element.c` includes `map_element.h`, which contains a declaration of `virtual_floor_paint`. I think there is a C idiosyncrasy  that I am missing.

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+
+void virtual_floor_paint(paint_session * session)
+{
+        // We only show when the placement modifier keys are active
```
Comments seem misaligned in this file

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+                    session->MapPosition.x <= tile->x + gMapVirtualFloorBaseSize &&
+                    session->MapPosition.y <= tile->y + gMapVirtualFloorBaseSize)
+                {
+                    show = true;
```
you can add `break;` here

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+                    show = true;
+                }
+            }
+            if (!show)
```
you can merge these two tail conditions into one after the `if` block starting in line 63

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+                session->MapPosition.x == tile->x &&
+                session->MapPosition.y == tile->y)
+            {
+                weAreLit = true;
```
`break`

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+            // Check if we are anywhere near the selection tiles (larger scenery / rides)
+        if (gMapSelectFlags & MAP_SELECT_FLAG_ENABLE_CONSTRUCT)
+        {
+            for (tile = gMapSelectionTiles; tile->x != -1; tile++)
```
I'd prefer if you sank the variable declarations to where they are used, declare `rct_xy16 * tile` in each of the loops that use it

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+            elementType == MAP_ELEMENT_TYPE_BANNER)
+        {
+            sint32 direction = map_element_get_direction(mapElement);
+            occupiedEdges   |= 1 << ((direction + get_current_rotation()) % 4);
```
cache the rotation, it won't change mid-function

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+                    if (theirLocationX == tile->x &&
+                        theirLocationY == tile->y)
+                    {
+                        theyAreLit = true;
```
`break`

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+
+        mapElement = map_get_first_element_at(theirLocationX >> 5, theirLocationY >> 5);
+        do
+        {
```
This whole loop and the condition above looks pretty much the same as the `we` variant of it. Any chance you could refactor them into their own function and just call it for different tiles?

janisozaur commented on src/openrct2/paint/map_element/virtual_floor.c:
```diff
@@ -0,0 +1,296 @@
+        sub_98197C(session, SPR_G2_SELECTION_EDGE_NE | (!(occupiedEdges & 0x1)? ((litEdges & 0x1)? remap_lit : remap_base) : remap_edge),
+            0, 0, 0, 0, 1, gMapVirtualFloorHeight,  5,  5, gMapVirtualFloorHeight + ((dullEdges & 0x1)? -2 : 0), get_current_rotation());
+    }
+    if (paintEdges & 0x2)
```
`0x2` seems out of order, does it also work if you move it above `0x1`?

janisozaur commented on src/openrct2/world/map.c:
```diff
@@ -4624,3 +4628,17 @@ uint16 check_max_allowable_land_rights_for_tile(uint8 x, uint8 y, uint8 base_z)
+}
+
+void map_invalidate_virtual_floor_tiles()
+{
```
empty? a placeholder for software renderer?

JeroenDStout commented:
I want to add a function call which removes (invalidates) the virtual floor when the modifiers are no longer used. In which function should I put this?

Gymnasiast commented:
@JeroenDStout Sounds like viewport.c to me, but I'd like to have that confirmed by @janisozaur .

JeroenDStout commented:
Lots of improvements, and it now is a bit more sane about checking whether the virtual floor needs to be drawn. Things still to do:

- Allow building paths to set the virtual floor height
- Add proper invalidation code to changes to the virtual floor (for software rendering)

JeroenDStout commented:
I have rebased and implemented a hopefully OK version of the invalidate tiles function. I unfortunately lack the time to consistently work on this, so if anybody wants to implement the missing two things, I would be rather happy.

1 - Any change in selection tiles should call `map_invalidate_virtual_floor_tiles` before and after for the invalidation to take effect. Specifically that would be when the player holds `ctrl` and moves the mouse around and when building a new part for a ride.  `map_invalidate_virtual_floor_tiles` invalidates based on the current selection tiles, so it needs to be called before *any* modification of them.

2 - While the footpath building window is open, every tick should have a call to `map_set_virtual_floor_height` with the corresponding footpath height so that footpath building will show the grid while holding a modifier key. If ride building or scenery building is active, that might have to take precedence (`ride.c` calls `scenery_tool_is_active()` to check this).

That should be it!

AaronVanGeffen commented:
I took the liberty to rebase this PR on top of develop. It seems to work nicely on OpenGL, but there seem to be some invalidation issues using the software renderer, still — as pointed out by @JeroenDStout in his last comment.

@IntelOrca or @LRFLEW, this PR seems to be nearly done. Would either of you be able to resolve the outstanding issues?

JeroenDStout commented:
Thank you, @AaronVanGeffen :)

janisozaur commented:
@AaronVanGeffen the CIs are failing because you are using a function provided by `openrct2-ui` in `libopenrct2`. That won't fly.

Which brings the issue: as this whole thing is really a UI-only thing, it should be a part of `openrct2-ui`

AaronVanGeffen commented:
Outstanding critical issues, as far as I can tell:

- [ ] Currently, `map.c` calls `input_test_place_object_modifier` in `map_set_virtual_floor_height`, which is unsupported in non-UI builds.
- [ ] Related: move the virtual floor into UI territory. Not necessarily in this PR — general selection hasn't been moved to the UI project yet, either.

I'd welcome input from others, with the above issues in particular. Other improvements to be made:

- [ ] Constructing a footpath does not yet support the virtual floor.
- [ ] When building a non-track ride (e.g. a Twist), the virtual floor is rendered on top of the ride, rather than below.

AaronVanGeffen commented:
I've once more rebased this PR on top of develop, and have hopefully addressed the issue with building the cli project. Some improvements to be made:

- [ ] Constructing a footpath does not yet support the virtual floor.
- [ ] When building a non-track ride (e.g. a Twist), the virtual floor is rendered on top of the ride, rather than below.

At some point, we should also move the virtual floor into UI territory. I don't believe that necessarily has to be done in this PR, though — general selection hasn't been moved to the UI project yet, either.

rwjuk commented on src/openrct2/windows/_legacy.c:
```diff
@@ -209,12 +209,15 @@ money32 place_provisional_track_piece(sint32 rideIndex, sint32 trackType, sint32
+            z_end = coords->z_end;
+        }
+        else
+            z_end = z_begin = coords->z_begin;
```
Can you wrap this in curly braces for clarity?

Gymnasiast replied:
+1. In general, either use braces on both the if and the else, or none at all.

rwjuk commented:
We also need to make sure that the virtual floor is invalidated when the selected element moves horizontally - at the moment it leaves artifacts behinds.

rwjuk commented on src/openrct2/world/footpath.c:
```diff
@@ -745,7 +745,22 @@ money32 footpath_provisional_set(sint32 type, sint32 x, sint32 y, sint32 z, sint
+        else
+        {
+            // Going up in the world!
+            map_set_virtual_floor_height((gFootpathProvisionalPosition.z + 1) * 8 + 4);
```
If we want the virtual floor to be at the top of the slope (like it's at the bottom when going down), this should be `map_set_virtual_floor_height((gFootpathProvisionalPosition.z + 2) * 8);`

AaronVanGeffen commented:
With footpath construction tackled as well, the following is left for this PR:
- [ ] Address proper horizontal invalidation, as per @rwjuk's last comment.
(Invalidate before _and_ after offsetting? We can come up with a less expensive solution, hopefully.)
- [ ] Test the code in all sorts of scenarios.
- [ ] Clean up commit history — stash a few (not all) commits together.

JeroenDStout commented:
Nice to see this progress!

A thing to note is that moving the floor vertically invalidates *all* involved tiles already. Moving horizontally without a clever buffer is in that way exactly as expensive as moving vertically, which I figured (at the time) to fall under 'optimise if it is noticeable' territory. I also considered that this is based on user input that only changes per-tile, so the slightly longer redraw time for those frames is acceptable in a 'non smooth' scenario.

Again, *really* nice to see this happening.

rwjuk commented:
I've tweaked the invalidation to correctly handle horizontal movement - @JeroenDStout, any improvements obviously welcome!

There seems to be a bug with the invalidation code in that it doesn't correctly take the 'do nothing' path when it could. The x and y components of `gVirtualFloorLastLocation` and  `min_position` seem to oscillate between the actual position and some other value (`32606` in the case I tested). Fixing that would improve performance.

AaronVanGeffen commented:
> [...] seem to oscillate between the actual position and some other value (32606 in the case I tested).

That sound suspiciously close to the limit of a `sint32`. Perhaps a `-1` losing its sign somewhere?

AaronVanGeffen commented:
I'm confident 4bfd8dc5309151cb0539608055b7200daa2602f3 has resolved all of the outstanding problems with invalidation in an optimal fashion. If that is indeed the case, all that's left is some thorough testing.

We should probably clean up the history by squashing a few commits before merging.

Gammadivided commented:
This sounds like a great idea, I just have one question: are you planning to implement this as an optional feature (as in, something you can turn on and off), and how hard would it be to make it optional?

AaronVanGeffen commented:
This is not currently planned to be optional. It wouldn't be that hard to make it optional, but making it optional would add an extra parameter to test, adding extra maintenance burden, of course…

Gammadivided commented:
Of course, I understand. I'd just be a bit bummed if this became a forced feature all of a sudden, since it could be distracting sometimes (not saying it isn't a useful feature).

AaronVanGeffen commented:
I talked with a few team members about it, and as a result it has been made optional as of 25d0431385e100e3e7674008bfc75675ac2b07b0. It will be on by default to make it more discoverable, though.

Gammadivided commented:
Much appreciated!

rwjuk commented:
![icarus park 2018-01-10 15-31-41](https://user-images.githubusercontent.com/16639257/34780586-70cb6b4e-f61b-11e7-925c-2b8fe346b379.png)

I'm assuming we're okay with the floor using the water palette? It'll mean the colour will be different in parks that use non-standard water colours. Hardly a big deal, but just FYI.

AaronVanGeffen commented:
Yes! Personally, I rather like it using the water palette. Makes for a really nice glassy touch, too.

Glad to hear the invalidation is working properly and performing well!

janisozaur commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -3140,6 +3140,13 @@ static void window_ride_construction_update_widgets(rct_window *w)
 static void window_ride_construction_select_map_tiles(Ride *ride, sint32 trackType, sint32 trackDirection, sint32 x, sint32 y)
 {
+        // If the scenery tool is active, we do not display our tiles as it
+        //  will conflict with larger scenery objects selecting tiles
```
misaligned

AaronVanGeffen replied:
Will fix it in a rebase. Thanks.

janisozaur commented on src/openrct2/paint/VirtualFloor.cpp:
```diff
@@ -0,0 +1,203 @@
+#include "../Input.h"
+#include "../interface/Viewport.h"
+#include "../sprites.h"
+#include "../world/Map.h"
```
these should be sorted

AaronVanGeffen replied:
Will fix it in a rebase. Thanks.

AaronVanGeffen commented:
> wasn't this supposed to be a part of UI?

Right now, it's very much tied into the selected tiles and painting routines, neither of which are in the UI subproject. I agree it would be better for all of that to be in the UI project, but that's beyond the scope of this PR, IMO.

AaronVanGeffen commented:
I've done what is hopefully the final rebase for this PR, squashing quite a few commits, bringing this PR down to 30 commits from 51 commits. Personally, I really like seeing how a feature has grown over time, so I think reducing it any more would do it a disservice. 😉

JeroenDStout commented:
Very nice work, @AaronVanGeffen ! I am very happy to see it. It all looks very good to me :)

janisozaur commented:
I briefly tested it yesterday and didn't spot any issues. The code looks good as well, so 👍 from me.
